x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java

```diff
@@ -301,28 +310,77 @@ && canAllocate(shard, allocation) == false) {
+                .toList();
+            Set<ShardRouting> unmovableShards = unmovableShardNodeDecisions.stream().map(e -> e.shard).collect(Collectors.toSet());
+
             long unmovableBytes = unmovableShards.stream()
                 .collect(Collectors.groupingBy(ShardRouting::currentNodeId))
                 .entrySet()
                 .stream()
                 .mapToLong(e -> unmovableSize(e.getKey(), e.getValue()))
                 .sum();
 
-            List<ShardRouting> unallocatedShards = candidates.stream()
+            List<ShardNodeDecisions> unallocatedShardNodeDecisions = candidates.stream()
                 .filter(Predicate.not(unmovableShards::contains))
-                .filter(s1 -> cannotAllocateDueToStorage(s1, allocation))
+                .flatMap(shard -> cannotAllocateDueToStorage(shard, allocation).stream())
                 .toList();
-            long unallocatableBytes = unallocatedShards.stream().mapToLong(this::sizeOf).sum();
+            List<ShardNodeDecisions> canRemainDecisionsForUnallocatedShards = unallocatedShardNodeDecisions.stream()
+                .filter(snd -> snd.nodeDecisions.size() > 0)
+                .map(ShardNodeDecisions::shard)
+                .map(
+                    shard -> new ShardNodeDecisions(
+                        shard,
+                        List.of(
+                            new NodeDecision(
+                                allocation.routingNodes().node(shard.currentNodeId()).node(),
+                                allocationDeciders.canRemain(shard, allocation.routingNodes().node(shard.currentNodeId()), allocation)
+                            )
+                        )
+                    )
+                )
+                .toList();
+            long unallocatableBytes = unallocatedShardNodeDecisions.stream().map(e -> e.shard).mapToLong(this::sizeOf).sum();
 
-            return new ShardsSize(
+            Map<ShardId, List<NodeDecision>> shardAllocateDecisions = Stream.concat(
+                unallocatedShardNodeDecisions.stream(),
+                otherNodesOnTierAllocationDecisions.stream()
+            )
+                .filter(shardNodeDecisions -> shardNodeDecisions.nodeDecisions.size() > 0)
+                .collect(toMap(snd -> snd.shard.shardId(), snd -> snd.nodeDecisions, (a, b) -> b));
+            return new ShardsAllocationResults(
                 unallocatableBytes + unmovableBytes,
-                Stream.concat(unmovableShards.stream(), unallocatedShards.stream())
-                    .map(ShardRouting::shardId)
-                    .collect(Collectors.toCollection(TreeSet::new))
+                Stream.concat(unmovableShardNodeDecisions.stream(), unallocatedShardNodeDecisions.stream())
+                    .map(e -> e.shard.shardId())
+                    .collect(Collectors.toCollection(TreeSet::new)),
+                Stream.concat(unmovableShardNodeDecisions.stream(), canRemainDecisionsForUnallocatedShards.stream())
```

Comment:
I think this can be simplified if we let `canRemainDecisionsForUnallocatedShards` and `otherNodesOnTierAllocationDecisions` contain both the remain and allocate decision, might need to convert to map directly. We can then combine the maps here at the end.

Reply:
I don't think that this can be simplified, unfortunately. `unallocatedShardNodeDecisions` and `unmovableShardNodeDecisions` need to be separate collections for the calculation of the allocated and unallocated bytes.
I also didn't understand the part about `both the remain and allocate decision`. I thought the point was that we track them separately in different collections.

Reply:
This patch 
```
Index: x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java b/x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java
--- a/x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java	(revision 740451bc3c1615fd5e627f2f633298c43d0103b5)
+++ b/x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java	(date 1667557190920)
@@ -286,7 +286,11 @@
             this.roles = roles;
         }
 
-        record ShardNodeDecisions(ShardRouting shard, List<NodeDecision> nodeDecisions) {}
+        record ShardNodeDecisions(ShardRouting shard, List<NodeDecision> nodeDecisions) {
+            static ShardNodeDecisions preferPrimary(ShardNodeDecisions snd1, ShardNodeDecisions snd2) {
+                return snd1.shard().primary() ? snd1 : snd2;
+            }
+        }
 
         public ShardsAllocationResults storagePreventsAllocation() {
             RoutingAllocation allocation = new RoutingAllocation(allocationDeciders, state, info, shardSizeInfo, System.nanoTime());
@@ -326,25 +330,17 @@
                 .filter(s -> allocatedToTier(s, allocation))
                 .flatMap(shard -> cannotRemainDueToStorage(shard, allocation).stream())
                 .toList();
-            List<ShardNodeDecisions> otherNodesOnTierAllocationDecisions = unmovableShardNodeDecisions.stream()
-                .map(ShardNodeDecisions::shard)
-                .map(
-                    shard -> new ShardNodeDecisions(
-                        shard,
-                        state.getRoutingNodes()
-                            .stream()
-                            .filter(n -> nodeTierPredicate.test(n.node()))
-                            .filter(n -> shard.currentNodeId().equals(n.nodeId()) == false)
-                            .map(
-                                n -> new NodeDecision(
-                                    n.node(),
-                                    withAllocationDebugEnabled(allocation, () -> allocationDeciders.canAllocate(shard, n, allocation))
-                                )
-                            )
-                            .toList()
+            Map<ShardId, ShardNodeDecisions> perShardIdUnmovable = unmovableShardNodeDecisions.stream()
+                .collect(toMap(snd -> snd.shard().shardId(), Function.identity(), ShardNodeDecisions::preferPrimary, TreeMap::new));
+            Map<ShardId, NodeDecisions> otherNodesOnTierAllocationDecisions = perShardIdUnmovable.values()
+                .stream()
+                .limit(MAX_AMOUNT_OF_SHARD_DECISIONS)
+                .collect(
+                    toMap(
+                        snd -> snd.shard().shardId(),
+                        snd -> new NodeDecisions(canAllocateDecisions(allocation, snd.shard()), snd.nodeDecisions())
                     )
-                )
-                .toList();
+                );
             Set<ShardRouting> unmovableShards = unmovableShardNodeDecisions.stream().map(e -> e.shard).collect(Collectors.toSet());
 
             long unmovableBytes = unmovableShards.stream()
@@ -358,53 +354,60 @@
                 .filter(Predicate.not(unmovableShards::contains))
                 .flatMap(shard -> cannotAllocateDueToStorage(shard, allocation).stream())
                 .toList();
-            List<ShardNodeDecisions> canRemainDecisionsForUnallocatedShards = unallocatedShardNodeDecisions.stream()
-                .filter(snd -> snd.nodeDecisions.size() > 0)
-                .map(ShardNodeDecisions::shard)
-                .map(
-                    shard -> new ShardNodeDecisions(
-                        shard,
-                        List.of(
-                            new NodeDecision(
-                                allocation.routingNodes().node(shard.currentNodeId()).node(),
-                                withAllocationDebugEnabled(
-                                    allocation,
-                                    () -> allocationDeciders.canRemain(
-                                        shard,
-                                        allocation.routingNodes().node(shard.currentNodeId()),
-                                        allocation
-                                    )
-                                )
-                            )
-                        )
-                    )
-                )
-                .toList();
+            Map<ShardId, ShardNodeDecisions> perShardIdUnallocated = unallocatedShardNodeDecisions.stream()
+                .collect(toMap(snd -> snd.shard().shardId(), Function.identity(), ShardNodeDecisions::preferPrimary, TreeMap::new));
+            Map<ShardId, NodeDecisions> canRemainDecisionsForUnallocatedShards = perShardIdUnallocated.values()
+                .stream()
+                .limit(MAX_AMOUNT_OF_SHARD_DECISIONS)
+                .collect(
+                    toMap(
+                        snd -> snd.shard().shardId(),
+                        snd -> new NodeDecisions(snd.nodeDecisions(), canRemainDecisions(allocation, snd.shard()))
+                    )
+                );
             long unallocatableBytes = unallocatedShardNodeDecisions.stream().map(e -> e.shard).mapToLong(this::sizeOf).sum();
 
-            Map<ShardId, List<NodeDecision>> shardAllocateDecisions = Stream.concat(
-                unallocatedShardNodeDecisions.stream(),
-                otherNodesOnTierAllocationDecisions.stream()
+            Map<ShardId, NodeDecisions> shardDecisions = Stream.concat(
+                canRemainDecisionsForUnallocatedShards.entrySet().stream(),
+                otherNodesOnTierAllocationDecisions.entrySet().stream()
             )
-                .filter(shardNodeDecisions -> shardNodeDecisions.nodeDecisions.size() > 0)
-                .collect(toMap(snd -> snd.shard.shardId(), snd -> snd.nodeDecisions, (a, b) -> b));
+                .collect(toSortedMap(Map.Entry::getKey, Map.Entry::getValue))
+                .entrySet()
+                .stream()
+                .limit(MAX_AMOUNT_OF_SHARD_DECISIONS)
+                .collect(toSortedMap(Map.Entry::getKey, Map.Entry::getValue));
             return new ShardsAllocationResults(
                 unallocatableBytes + unmovableBytes,
                 Stream.concat(unmovableShardNodeDecisions.stream(), unallocatedShardNodeDecisions.stream())
                     .map(e -> e.shard.shardId())
                     .collect(Collectors.toCollection(TreeSet::new)),
-                Stream.concat(unmovableShardNodeDecisions.stream(), canRemainDecisionsForUnallocatedShards.stream())
-                    .filter(shardNodeDecisions -> shardNodeDecisions.nodeDecisions.size() > 0)
-                    .collect(
-                        toSortedMap(
-                            snd -> snd.shard.shardId(),
-                            snd -> new NodeDecisions(shardAllocateDecisions.getOrDefault(snd.shard.shardId(), List.of()), snd.nodeDecisions)
-                        )
-                    )
-                    .entrySet()
-                    .stream()
-                    .limit(MAX_AMOUNT_OF_SHARD_DECISIONS)
-                    .collect(toSortedMap(Map.Entry::getKey, Map.Entry::getValue))
+                shardDecisions
+            );
+        }
+
+        private List<NodeDecision> canAllocateDecisions(RoutingAllocation allocation, ShardRouting shardRouting) {
+            return state.getRoutingNodes()
+                .stream()
+                .filter(n -> nodeTierPredicate.test(n.node()))
+                .filter(n -> shardRouting.currentNodeId().equals(n.nodeId()) == false)
+                .map(
+                    n -> new NodeDecision(
+                        n.node(),
+                        withAllocationDebugEnabled(allocation, () -> allocationDeciders.canAllocate(shardRouting, n, allocation))
+                    )
+                )
+                .toList();
+        }
+
+        private List<NodeDecision> canRemainDecisions(RoutingAllocation allocation, ShardRouting shard) {
+            return List.of(
+                new NodeDecision(
+                    allocation.routingNodes().node(shard.currentNodeId()).node(),
+                    withAllocationDebugEnabled(
+                        allocation,
+                        () -> allocationDeciders.canRemain(shard, allocation.routingNodes().node(shard.currentNodeId()), allocation)
+                    )
+                )
             );
         } 
```

demonstrates what I mean. First of all, we can limit the number of decisions we collect earlier. I also ensured we prefer primary over replica decisions. And removed the filter for empty node decisions list (new tier, see other comment). The latter requires a minor change to `testMoveToEmpty`.